src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs

```diff
@@ -62,68 +63,172 @@ public void ScopeDisposed(int serviceProviderHashCode, int scopedServicesResolve
+            Justification = "Parameters to this method are primitive and are trimmer safe.")]
+        [Event(8, Level = EventLevel.Verbose)]
+        private void ServiceProviderDescriptors(int serviceProviderHashCode, string descriptors, int chunkIndex, int chunkCount)
+        {
+            WriteEvent(8, serviceProviderHashCode, descriptors, chunkIndex, chunkCount);
         }
 
         [NonEvent]
-        public void ServiceResolved(Type serviceType)
+        public void ServiceResolved(ServiceProvider provider, Type serviceType)
         {
             if (IsEnabled(EventLevel.Verbose, EventKeywords.All))
             {
-                ServiceResolved(serviceType.ToString());
+                ServiceResolved(serviceType.ToString(), provider.GetHashCode());
             }
         }
 
         [NonEvent]
-        public void CallSiteBuilt(Type serviceType, ServiceCallSite callSite)
+        public void CallSiteBuilt(ServiceProvider provider, Type serviceType, ServiceCallSite callSite)
         {
             if (IsEnabled(EventLevel.Verbose, EventKeywords.All))
             {
                 string format = CallSiteJsonFormatter.Instance.Format(callSite);
                 int chunkCount = format.Length / MaxChunkSize + (format.Length % MaxChunkSize > 0 ? 1 : 0);
 
+                int providerHashCode = provider.GetHashCode();
                 for (int i = 0; i < chunkCount; i++)
                 {
                     CallSiteBuilt(
                         serviceType.ToString(),
-                        format.Substring(i * MaxChunkSize, Math.Min(MaxChunkSize, format.Length - i * MaxChunkSize)), i, chunkCount);
+                        format.Substring(i * MaxChunkSize, Math.Min(MaxChunkSize, format.Length - i * MaxChunkSize)), i, chunkCount,
+                        providerHashCode);
                 }
             }
         }
 
         [NonEvent]
-        public void DynamicMethodBuilt(Type serviceType, int methodSize)
+        public void DynamicMethodBuilt(ServiceProvider provider, Type serviceType, int methodSize)
         {
             if (IsEnabled(EventLevel.Verbose, EventKeywords.All))
             {
-                DynamicMethodBuilt(serviceType.ToString(), methodSize);
+                DynamicMethodBuilt(serviceType.ToString(), methodSize, provider.GetHashCode());
             }
         }
 
         [NonEvent]
-        public void ServiceRealizationFailed(Exception exception)
+        public void ServiceRealizationFailed(Exception exception, int serviceProviderHashCode)
         {
             if (IsEnabled(EventLevel.Error, EventKeywords.All))
             {
-                ServiceRealizationFailed(exception.ToString());
+                ServiceRealizationFailed(exception.ToString(), serviceProviderHashCode);
             }
         }
+
+        [NonEvent]
+        public void ServiceProviderBuilt(ServiceProvider provider)
+        {
+            if (IsEnabled(EventLevel.Verbose, EventKeywords.All))
```

Comment:
I'd like to be able to have this fire OnEventCommand if I use a specific keyword so we can get access to this data even if the application is already built.

Here's what I did in Kestrel https://github.com/dotnet/aspnetcore/blob/545d4eece7fcc5a67b27f6ba96636d9f476741ae/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs#L354. The idea here is to handle when the event source is enabled and then fire the event (if the given keyword is on). In the kestrel case it always fires the event today because it doesn't have that much data (maybe we should tweak this), but for DI, we'd need to capture *all* service providers built in the application and then replay them here (and probably drop them from being rooted from the event source on dispose).